### PR TITLE
generate roadmap in json format & deploy to Pages

### PR DESCRIPTION
**Pull request by mordamax**

### CODE REVIEW

File: src/fetchProjectItems.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                      }
+                      break
+                    case "ProjectV2ItemFieldNumberValue":
+                      if (fv.node?.field.__typename === "ProjectV2Field") {
+                        projectItem.content.customFields[fv.node.field.name] = fv.node.number
+                      }
+                      break
+                  }
+                })
+            }
```
Comment (mutantcornholio) [inline src/fetchProjectItems.ts:62]: I can't understand what's going on here. Could you simplify/split/comment it somehow?

File: src/main.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        console.log(`Created ${OUTPUT_FILE} 👍`)
+      })
+    } catch (e) {
+      console.error(e)
+    }
+  } catch (e) {
+    console.error(e)
+  }
+
+})().catch(console.error)
```
Comment (mutantcornholio) [inline src/main.ts:25]: What's the reason behind all these catches? Why do we need the code not to do anything, yet not fail?

**Review by mutantcornholio — CHANGES_REQUESTED** [review]

File: src/main.ts (hunk 2 above)
Comment (mordamax) [reply src/main.ts:25]: @mutantcornholio it was failing silently without any logs and 0 result, so i added these, but I remember during iterations of refactoring there was a mistake on how it was handled, so may be that was a reason.
So you propose just remove all try/catches and let it fail with errors in console? that sounds right to me

Comment (mutantcornholio) [reply src/main.ts:25]: Considering that you're handling async code, I propose something like this:
```ts
(async () => {
    const project = await fetchProjectItems({ org, projectId, first: 100 })
    await fs.mkdir("./dist")
    await fs.writeFile(OUTPUT_FILE, JSON.stringify(project))
    console.log(`Created ${OUTPUT_FILE} 👍`)
})().catch(e => {
    console.error(e);
    process.exit(1);
})
```

File: src/main.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { fetchProjectItems } from "./fetchProjectItems";
+import { config } from "./config";
+import * as fs from "fs";
```
Comment (mutantcornholio) [inline src/main.ts:3]: Considering that you're `await`ing on it... shouldn't it be 
```ts
import { promises as fs } from "fs"
```
?

File: src/fetchProjectItems.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import { print } from "graphql/index";
+import { octokit } from "./octokit";
+
+export async function fetchProjectItems(
+  params: GetProjectsQueryVariables,
+  recursiveCollection?: ProjectDetails
+): Promise<ProjectDetails> {
+  console.log(`Fetching projects`, params)
+  // @ts-ignore
+  const project: GetProjectsQuery = await fetchProject(params).catch(console.error)
```
Comment (dvdplm) [inline src/fetchProjectItems.ts:17]: Isn't this swallowing the error? That is a bit unfortunate. Seems better to either crash or return the error.

File: src/fetchProjectItems.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  console.log(project)
+
+  if (project.organization?.projectV2?.items?.edges) {
+    const { hasNextPage, endCursor } = project.organization?.projectV2?.items?.pageInfo
+    const items: ProjectItem[] = [];
+
+    project.organization?.projectV2?.items?.edges
+      ?.map(item => item?.node)
+      ?.map(item => {
+        if (item) {
```
Comment (dvdplm) [inline src/fetchProjectItems.ts]: `item` is a `node` here right? Maybe rename it?

File: src/fetchProjectItems.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      ? await fetchProjectItems({ ...params, after: endCursor }, recursiveCollection)
+      : recursiveCollection
+  } else {
+    throw new Error(`No items found in a project ¯\_(ツ)_/¯; params: ${JSON.stringify(params)}`)
+  }
+}
+
+async function fetchProject(params: GetProjectsQueryVariables) {
+  return await octokit.graphql<GetProjectsQuery>(print(GetProjectsDocument), params)
+    .catch(console.error)
```
Comment (dvdplm) [inline src/fetchProjectItems.ts:96]: Here too: isn't it better to return the error and handle it upstream in the call graph instead?

File: src/main.ts (hunk 3 above)
Comment (mordamax) [reply src/main.ts]: [1970a31](https://github.com/paritytech/roadmap/pull/10/commits/1970a311259a87f2f52d8d330443b1fa13c289fe)

File: src/main.ts (hunk 2 above)
Comment (mordamax) [reply src/main.ts]: [1970a31](https://github.com/paritytech/roadmap/pull/10/commits/1970a311259a87f2f52d8d330443b1fa13c289fe)

File: src/types.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import {GetProjectsQuery, IssueState, IssueStateReason} from "./generated/graphql-operations";
```
Comment (dvdplm) [inline src/types.ts:1]: `graphql-operations` is a pretty big file; just out of curiosity: do we need the whole file?

File: src/fetchProjectItems.ts (hunk 1 above)
Comment (mordamax) [reply src/fetchProjectItems.ts]: I see why that looks complicated
I tried to simplify a bit, added comments what exactly going on and reasoning and moved out the label extraction.
However It was harder to split/simplify the rest of the fields extractions due to the fact of GraphQL + Typings, generate custom type accordingly to what i have queried, and it doesn't unfortunately split its nested types, so I could use these NestedTypeName in my separate method. The only way is to copy some part of generated type into a argument type definition but it's toooo long and really complicates further support. 
As if you add more fields, you'll have to add them in these definitions as well
image
like this 
image

I know it's not ideal, but i hope comments will simplify support in future

File: src/config.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+dotenv.config()
+
+const auth = process.env.ACCESS_TOKEN
+const org = process.env.GITHUB_ORG;
+const projectId: number = parseInt(process.env.PROJECT_ID!, 10);
+
+if (!projectId || isNaN(projectId)) throw new Error('PROJECT_ID must be provided')
+if (!org) throw new Error('GITHUB_ORG must be provided')
+if (!auth) throw new Error('GITHUB_TOKEN must be provided')
```
[the remark this answers is not in the document]
Comment (mordamax) [reply src/config.ts:11]: i thought it'd be enough, but may be you something more specific in mind?

[the remark this answers is not in the document]
Comment (dvdplm) [reply src/config.ts:11]: My thinking went "if we're hacked and the env vars changed, how bad would it be?", but I guess the damage would be rather limited. It's ok as-is.

File: src/types.ts (hunk 7 above)
Comment (mordamax) [reply src/types.ts:1]: yep, this whole github graphql schema mapped to TypeScript. highly improves support as enables static typing of graphql responses based on default schema + your custom queries. 
For example GetProjectsQuery type is built on top of this query https://github.com/paritytech/roadmap/blob/664732a2830a01ef1283c85bbb89e0c3407e2c75/getProjects.graphql
so when I work with response, it has only those fields that I requested. But yes, we need whole file

we can either ignore this file or keep in git, i preferred to keep

File: src/fetchProjectItems.ts (hunk 1 above)
Comment (mutantcornholio) [reply src/fetchProjectItems.ts]: Comments do help, thank you

File: src/main.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+const { projectId, org } = config;
+
+(async () => {
+  try {
+    const project = await fetchProjectItems({ org, projectId, first: 100 })
+    console.log(project)
+    try {
+      await fs.mkdir('./dist', () => {
+        fs.writeFileSync(OUTPUT_FILE, JSON.stringify(project), )
+        console.log(`Created ${OUTPUT_FILE} 👍`)
```
[the remark this answers is not in the document]
Comment (mutantcornholio) [reply src/main.ts]: I disagree. The operations that are more likely to fail should be called first, and network IO is more likely to fail, right?

[the remark this answers is not in the document]
Comment (mordamax) [reply src/main.ts]: i think this case would be more actual if the environment would be unknown and more dynamic. Here we run it always through the GHA, which is more stable, so if it has no problem with creation now, then unlikely later, nevertheless if it fails after operation was made - not a problem here

File: src/types.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    state: IssueState,
+    stateReason?: IssueStateReason | null,
+    closed: boolean,
+    closedAt?: any | null,
+    updatedAt: any,
+    publishedAt?: any | null,
+    author?: { avatarUrl: any, login: string, url: any } | null
+    labels?: { name: string, url: string, color: string }[] | undefined
+    customFields: ProjectCustomFields
+  }
```
[the remark this answers is not in the document]
Comment (mordamax) [reply src/types.ts]: [f553ef0](https://github.com/paritytech/roadmap/pull/10/commits/f553ef0f964f6c2d5d7b3782342afc2ab5b9025c)

File: src/fetchProjectItems.ts (hunk 6 above)
Comment (mordamax) [reply src/fetchProjectItems.ts]: [f52d86d](https://github.com/paritytech/roadmap/pull/10/commits/f52d86da25ecb5114aab62186378ad78e85710ee)

File: src/fetchProjectItems.ts (hunk 5 above)
Comment (mordamax) [reply src/fetchProjectItems.ts]: [f52d86d](https://github.com/paritytech/roadmap/pull/10/commits/f52d86da25ecb5114aab62186378ad78e85710ee)

File: src/fetchProjectItems.ts (hunk 4 above)
Comment (mordamax) [reply src/fetchProjectItems.ts]: [f52d86d](https://github.com/paritytech/roadmap/pull/10/commits/f52d86da25ecb5114aab62186378ad78e85710ee)

File: src/config.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import * as dotenv from "dotenv";
+
+dotenv.config()
+
+const auth = process.env.ACCESS_TOKEN
+const org = process.env.GITHUB_ORG;
+const projectId: number = parseInt(process.env.PROJECT_ID!, 10);
+
+if (!projectId || isNaN(projectId)) throw new Error('PROJECT_ID must be provided')
```
[the remark this answers is not in the document]
Comment (mordamax) [reply src/config.ts]: added README and also comment here 
f553ef0

**Review by mutantcornholio — APPROVED** [review]

File: src/fetchProjectItems.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+import { print } from "graphql/index";
+import { octokit } from "./octokit";
+
+export async function fetchProjectItems(
+  params: GetProjectsQueryVariables,
+  recursiveCollection?: ProjectDetails
+): Promise<ProjectDetails> {
+  console.log(`Fetching projects`, params)
+  // @ts-ignore
+  const project: GetProjectsQuery = await fetchProject(params).catch((e) => {
```
Comment (dvdplm) [inline src/fetchProjectItems.ts:16]: On line 100 the `fetchProject` implementation already logs&exits, so is this catch actually needed?

**Review by dvdplm — APPROVED** [review]
Lingering doubts about `catch` best practices but overall lgtm.

**Review by the-right-joyce — APPROVED** [review]
lgtm

**Pull request merged**